### PR TITLE
Split out action tests

### PR DESCRIPTION
**Pull request by mhluska**
I think this library would benefit from splitting up the source and test code into multiple files. The main benefit to doing it this way is it's easier to tell at a glance what the dependencies are for each chunk of code.

I've done it here for one section: the get action tests. @mrichar1 if you like where I'm going with this, I can finish the work for the remaining tests. Let me know.

The idea is to move the fixtures into their own files where a function generates them from scratch each time. And then `require` the relevant describes as necessary from the top level test file.

I've also added some spacing around each unit test for readability.

I think we could benefit from running eslint on the tests in the future too.

### CODE REVIEW

File: tests/unit/jsonapi-vuex.spec.js [hunk 1]
```diff
@@ -326,6 +258,21 @@ beforeEach(function() {
     }
   }
 
+  // Variables Used by required submodules.
+  // HACK: Rely on `this` so that submodules can access them. Eventually, we
+  // will want to define these in fixture files and import them as needed.
+  // Could use a factory library. See `fixtures/json_index_1.js`.
+  this.api = api;
```
Comment (mhluska) [inline tests/unit/jsonapi-vuex.spec.js]: This will all go away if we continue with splitting up the tests into multiple files.

**Comment by mrichar1** [comment]
Yes, I'm happy for the unit tests to get refactored in this way - if you're happy to keep working on this then by all means extend this PR to be a complete refactor, rather than having to hack it so that some parts are 'external' and others 'internal'.

A few thoughts:

1. I've updated `README.md` to have a section on 'Developement' to highlight how the tests are expected to work, some coding policy etc.

2. `eslint` is configured for both the main code and tests, and is launched via `karma-webpack` when the tests run. At present in some cases it reports errors/warnings but then continues to run the tests anyway - I need to try to make it 'fail hard'.

3. The `machine_X` data should probably also be fixtures, rather than explicitly defined in `get.spec` (even if they're only imported in `get.spec`).

4. I wonder if it might be better to keep some of the fixtures together - for example, keeping the `json`, `norm` and `store` versions of e.g. `widget_3` together in one file? you could still export them separately, and get some flexibility as to use, e.g:

fixtures/widget_3.js
```
const json = () => { ... }console.log(json())
const norm = () => { ... }
const store = () => { ... }

export { json, norm, store }
```

test.js
```
import { json } from './fixtures/widget_3'
console.log(json())

// Or

import { json as json_widget_3 } from './fixtures/widget_3'
console.log(json_widget_3())


// Or 

import * as widget_3 from './fixtures/widget_3'
console.log(widget_3.json())
```

Let me know what you think!

**Comment by mhluska** [comment]
That works for me. I'll finish this PR and put the json, norm and store fixtures together.

**Comment by mrichar1** [comment]
Great! I look forward to seeing the full PR in due course.

The variable naming in the project is consistent, but it's come from an in-house style guide, which is that functions are camelCase, while variables are underscore. I'm happy to switch away from this if it's a confusing standard (I certainly don't think it's a well-used one!) - I have a very slight preference for camelCase but am flexible.

**Comment by mhluska** [comment]
I'd be fine moving to all camelCase in another PR down the road. I guess I'm fine with any style as long as we're consistent. Granted I am randomly mixing underscore and camel case in this PR 😛 

So I've finished the majority of the work for the get actions. I wanted your thoughts again before I continue.

The main thing you might be wondering about is why am I sometimes using `excludingEvery('_jv')`? The reason is we lost the self-referential _jv key by moving to completely decoupled fixtures. Let me give an example:

```js
export function normFormatWithRels(widget) {
  const widget = normFormat()
  widget._jv.rels = { widgets: widget }
  return widget
}
```

This function uses an existing widget passed in from the outside and looks like this:

```
{ '1':
   { foo: 1,
     bar: 'baz',
     _jv:
      { type: 'widget',
        id: '1',
        relationships:
         { widgets:
            { data: { type: 'widget', id: '2' },
              links: { related: '/widget/1/widgets' } } },
        rels:
         { widgets:
            { foo: 2,
              _jv:
               { type: 'widget',
                 id: '2',
                 relationships: { widgets: { data: [ { type: 'widget', id: '1' }, { type: 'widget', id: '3' } ] } },
                 rels:
                  { widgets:
                     { '1': { foo: 1, bar: 'baz', _jv: [Circular] },
                       '3':
                        { foo: 3,
                          _jv:
                           { type: 'widget',
                             id: '3',
                             relationships: { widgets: { data: { type: 'widget', id: '1' } } },
                             rels: { widgets: { foo: 1, bar: 'baz', _jv: [Circular] } } } } } } } } } } },
```

However, to keep things nice and decoupled, we'd prefer if the function just generates its own rels value:

```js
export function normFormatWithRels() {
  const widget = normFormat()
  widget._jv.rels = { widgets: createNormWidget2() }
  return widget
}
```

The issue is the shape of the data is now missing the circular reference:

```
{ foo: 1,
     bar: 'baz',
     _jv:
      { type: 'widget',
        id: '1',
        relationships:
         { widgets:
            { data: { type: 'widget', id: '2' },
              links: { related: '/widget/1/widgets' } } },
        rels:
         { widgets:
            { foo: 2,
              _jv:
               { type: 'widget',
                 id: '2',
                 relationships: { widgets: { data: [ { type: 'widget', id: '1' }, { type: 'widget', id: '3' } ] } } } } } } },
```

So in the tests I had to sometimes exclude comparison on the `_jv` key.

**Comment by mhluska** [comment]
This PR should be good to go.

**Comment by mrichar1** [comment]
Thanks for this - it looks great! It seems to be the 'right amount' of splitting out - and yes, 400 more lines, but since that is only in the test suite, and all in the aid of clarity and separation it's fine with me!

The circular reference stuff is interesting - I *thought* I'd got them all when I first reviewed the code a while ago - hence the presence of all the `clone` calls in various places - but it's very likely that I missed some (my main language is python, so diving deep into JS has been an experience with everything being a reference, and all the async stuff! :rofl: ) It's good to refactor things like this and see what creeps out of the woodwork!

Having the fixtures call other fixtures for sub-data is definitely the right direction to head in...  if nothing else calling a function guarantees a 'copy' of the data, so we don't risk passing references and modifying things unexpectedly in the tests.

There's a lot going on just now so I'll leave off camelCase until it's quiet then do an atomic commit - but I'll open an issue to remind me to do this!

Thanks once again for all this work!

**Pull request merged**